Feature/issue 253 amazonmq broker user password

tmcelhattan opened the pull request:
Adding custom rules and tests for issue #253 - AWS::AmazonMQ::Broker User Password property rules.

erickascic commented on lib/cfn-nag/custom_rules/AmazonMQBrokerUserPasswordRule.rb:
```diff
@@ -0,0 +1,40 @@
+    'F52'
+  end
+
+  def resource_type
```
this method is unnecessary since this class doesn't derive from PasswordBaseRule

erickascic commented on lib/cfn-nag/custom_rules/AmazonMQBrokerUserPasswordRule.rb:
```diff
@@ -0,0 +1,40 @@
+
+    violating_resources = resources.select do |resource|
+      violating_users = resource.users.select do |user|
+        insecure_parameter?(cfn_model, user['Password']) || \
```
1. break this logic out into a predicate.  the backslashes give me the willies, but aside from that a complex boolean like this deserves its own named method to lower CCM and improve readability

2. checking the nil.... i'm hemming and hawing over that.  cfn-model is the underlying gem that is supposed to parse and prepare things so that we don't have to do a lot of this checking in nag rules.  since this is a nested data structure though (User) we are stuck doing some of that work here.  i guess i'd suggest checking it's there before trying to reference it... but then if it's not there, just do nothing.

erickascic commented on spec/custom_rules/AmazonMQBrokerUserPasswordRule_spec.rb:
```diff
@@ -0,0 +1,101 @@
+    end
+  end
+
+  context 'AmazonMQBroker has password from Not Secure String in Systems Manager' do
```
missing password is one of the contexts to capture for coverage

erickascic commented on lib/cfn-nag/custom_rules/AmazonMQBrokerUserPasswordRule.rb:
```diff
@@ -0,0 +1,40 @@
+      violating_users = resource.users.select do |user|
+        insecure_parameter?(cfn_model, user['Password']) || \
+        insecure_string_or_dynamic_reference?(cfn_model, user['Password']) || \
+        user['Password'].nil?
```
actually.... can't you derive form PasswordBaseRule and use the subproperty behavior?  i'm not for sure on that

erickascic commented on lib/cfn-nag/custom_rules/AmazonMQBrokerUserPasswordRule.rb:
```diff
@@ -0,0 +1,67 @@
+    end
+  end
+
+  def get_violating_users(cfn_model, resource)
```
call this violating_users?  name the resource argument as"mq_broker

erickascic replied:
make this private

erickascic commented on lib/cfn-nag/custom_rules/AmazonMQBrokerUserPasswordRule.rb:
```diff
@@ -0,0 +1,67 @@
+    end
+  end
+
+  def check_user_property(user_property)
```
i'd nuke this method.  the return values could be true or nil which is possibly confusing - but you can check the nil? inline where this is invoked imho

erickascic replied:
make this private

erickascic commented on lib/cfn-nag/custom_rules/AmazonMQBrokerUserPasswordRule.rb:
```diff
@@ -0,0 +1,67 @@
+    'F52'
+  end
+
+  def check_password_property(cfn_model, user)
```
'check' is a bit opaque as a term.  while i would never say comments are never necessary, imho writing them usually means something isn't obvious from the code itself.

user_has_insecure_password? is probably a better name

tmcelhattan commented:
@erickascic Made some additional refactors based on your comments:
1. Removed that `check_user_property` method
2. Renamed the methods to make what they perform more clear
3. Made those renamed class methods private.
